builtins: stream consistency checker output

tbg opened the pull request:
Also makes it resilient to per-Range errors, which now no longer
tank the entire operation.

```sql
-- avoid buffering in cli
\set display_format=csv;
-- avoid rows getting buffered at server
set avoid_buffering=true;
-- compute as fast as possible
SET CLUSTER SETTING server.consistency_check.max_rate = '1tb';

SELECT * FROM crdb_internal.check_consistency(false, '', '');
```

edit: should probably use `SET CLUSTER SETTING sql.defaults.results_buffer.size = '1b';` here, somehow the `set avoid_buffering=true` isn't cutting it entirely

Release justification: improvement for a debugging-related feature
Release note: None

tbg committed:
builtins: stream consistency checker output

Also makes it resilient to per-Range errors, which now no longer
tank the entire operation.

```sql
-- avoid buffering in cli
\set display_format=csv;
-- avoid rows getting buffered at server
set avoid_buffering=true;
-- compute as fast as possible
SET CLUSTER SETTING server.consistency_check.max_rate = '1tb';

SELECT * FROM crdb_internal.check_consistency(false, '', '');
```

Release justification: improvement for a debugging-related feature
Release note: None

tbg committed:
builtins: add `duration` column to check_consistency

Noticed that some checks were taking a very long time but it's hard to
tell which ones unless setting up and watching streaming output.

Add a duration column. This can also help us reason about the general
speed of the consistency checker; it does do very expensive hashing
at the moment (sha256) when something much cheaper and weaker would
do.

Release justification: debug improvement for an internal builtin
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/87378)
<!-- Reviewable:end -->

tbg commented:
NB: probably need to tweak something in the example above to truly disable buffering, see [^1]

[^1]: https://cockroachlabs.slack.com/archives/C0168LW5THS/p1661984789936559

pavelkalinnikov commented on pkg/sql/sem/builtins/generator_builtins.go:
```diff
@@ -1891,18 +1899,62 @@ func (*checkConsistencyGenerator) ResolvedType() *types.T {
+		// the result set.
+		desc := c.descs[0]
+		c.descs = c.descs[1:]
+		resp, err := c.consistencyChecker.CheckConsistency(
```
How does this request behave if the start and end key get out of sync with the range descriptor?

Also, originally, how was this request decomposed into per-range requests? Is it by some magic of `DistSender`?

Next to the magic that previously decomposed this request, is there a corresponding magic that streams the response instead of blocking on it fully? Can we push this behaviour as low as that level, so that we don't need to manually deal with meta ranges up here?

pavelkalinnikov commented:
I wonder if this PR eliminates the need for the "special code" mentioned here:
https://github.com/cockroachdb/cockroach/blob/544c6ae60c6259fd52087e417bd548271bc771ac/pkg/sql/consistencychecker/consistency_checker.go#L47-L48

tbg reviewed (commented):
> I wonder if this PR eliminates the need for the "special code" mentioned here:

In practice yes, but I don't want to back it out to account for cases where the generator reads a range as `[a,z)` but a rapid sequence of splits cuts it into `[a,b)`, `[b,c)`, ..., `[y,z)`. In that case, we don't want DistSender to send dozens of parallel requests.

TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-:-NBNXYcWAtOeqQr0fBaO:b-ehk7nu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov)

___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1903 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBCZLz9YeiK8FX4HtAC-r1-1903:-NBNWSh-CHN6a1M3y0ur:b-67ywoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/34545467a661cf2152620d395ba88b71376860a7/pkg/sql/sem/builtins/generator_builtins.go#L1903)):*
Done. I don't think wrapping at 80 is worth review feedback in general, though. I do realize it's in our guidelines but I think that is broadly ill-advised. If wrapping is important, gofmt needs to do it; otherwise we should be lenient to reduce drag. (These are my opinions and there's probably someone out there who feels strongly the other way). I feel the same about harmless typos, punctuation, etc, though I have a harder time myself not pointing them out.

> Will you do it in this PR? If not, add a TODO / link to an issue?

Good idea, filed this as a starter project for KV: https://github.com/cockroachdb/cockroach/issues/87503
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1942 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBCZLz9YeiK8FX4HtAC-r1-1942:-NBNX-dtAVjDrt_N3_gn:bv5ojz7) ([raw file](https://github.com/cockroachdb/cockroach/blob/34545467a661cf2152620d395ba88b71376860a7/pkg/sql/sem/builtins/generator_builtins.go#L1942)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

How does this request behave if the start and end key get out of sync in presence of range splits/merges? Or, since it's in a Txn, we are fine?

Also, originally, how was this request decomposed into per-range requests? Is it by some magic of `DistSender`?

It looks like `DistSender` can't stream the result, hence we are here? But can we push this behaviour as low as that level, so that we don't need to manually deal with meta ranges up here?
</blockquote></details>

DistSender doesn't support streaming (fixing this is a deep change, see `kvstreamer`, where we preferred bolting on vs rewriting DistSender), so cutting the request into pieces here is a pragmatic way to get it done. DistSender can split and recombine `CheckConsistency` though, so if we get it wrong here it'll just do that for us, so this will work even if the range boundaries change out from under us. What matters is that we read a consistent snapshot of the range descriptors, so we're going to cover the entire keyspace. I think we could get duplicate results if we have a range `[a,c)` and `[c,d)` but it really merged into `[a,d)` - in that case, we'd ask DistSender to scan the same range twice and I think it would produce two consistency checks for `[a,d)` in return. That is acceptable.


<!-- Sent from Reviewable.io -->

tbg committed:
builtins: refactor check_consistency for clarity

This implements a number of suggestions made by Pavel[^1].

[^1]: https://github.com/cockroachdb/cockroach/pull/87378#discussion_r965081891

Release justification: None needed; 22.2 has been cut
Release note: None

tbg commented on pkg/sql/sem/builtins/generator_builtins.go:
```diff
@@ -1891,18 +1900,62 @@ func (*checkConsistencyGenerator) ResolvedType() *types.T {
 
 // Next is part of the tree.ValueGenerator interface.
-func (c *checkConsistencyGenerator) Next(_ context.Context) (bool, error) {
+func (c *checkConsistencyGenerator) Next(ctx context.Context) (bool, error) {
```
Good idea, PTAL

pav-kv reviewed (approved):
Mostly style nits. Address as many as you like :)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-:-NBTSyNp2gTuUKy-C48S:b8c9a26)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov and @tbg)

___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1837 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBT-2ATBG3DjrJHt--9:-NBT-2ATBG3DjrJHt--A:b2bmlrk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1837)):*
> ```Go
> 
> 	descs []roachpb.RangeDescriptor
> 	// remainingRows is populated by Next() call peels of the first
> ```

This sentence got corrupted, and there is no longer a `curRow` field mentioned in it, please fix.
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1840 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBTANX8-8_SGPO5rj_d:-NBTANX8-8_SGPO5rj_e:b7irbst) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1840)):*
> ```Go
> 	// row and moves it to curRow. When empty, consumes from 'descs' to produce
> 	// more rows.
> 	remainingRows      []roachpb.CheckConsistencyResponse_Result
> ```

style nit / promoted content:

IMO in Go brevity is a feature, e.g. in naming. I would name this as `rows` or `resps`/`results`, for example. The "remaining" bit can be inferred from code/comments, and, besides, `remainingRows` now represents only a prefix of remaining rows to be fully precise (as opposed to, previously, the full set of them), so it's better to make the name purposefully vague than not exactly precise.

Same about the duration field below. Something like `lastDur`/`requestDur`/`refillDur` would do.

[pre-existing] `consistencyChecker` (note how it stands out from other short field names) I would name as `checker`/`runner`, if not `ch` or `cc`, or just embed it, as both "consistency" and "check" are inferred from its type and the wrapping type, plus this file is all about consistency checking anyway.
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1917 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBTMqBN1unDcRTXMVwu:-NBTMqBN1unDcRTXMVwv:b8de1m6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1917)):*
> ```Go
> 			return err
> 		}
> 		if len(desc.StartKey) == 0 {
> ```

I don't think I follow this bit, but that's probably due to my limited knowledge about meta ranges for now. I trust this bit works, but just asking questions to understand it. If you think there is a simple/short way to reflect it in the code comments, feel free (e.g. link to some doc / package / file).

I gather that `meta1`/`meta2` are levels of the 2-level "tree" of meta ranges (i.e. ranges that tell us what/where all ranges are). When the whole keyspace is small, we only have one meta1, and (by the looks of it from the comment) also one meta2 (as opposed to not having meta2 at all). Correct?

Can both meta1 and meta2 split, or one of them is always a singleton? It both can split, where is the "global" root of things, i.e. is there some meta0 too?

What is `r1`, and why can we get a second copy?



_Code quote:_
```Go
		if len(desc.StartKey) == 0 {
			desc.StartKey = keys.MustAddr(keys.LocalMax)
			// Elide potential second copy we might be getting for r1
			// if meta1 and meta2 haven't split.
			// This too should no longer be necessary with IterateRangeDescriptors.
			if len(c.descs) == 1 {
				continue
			}
		}
```
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1939 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBT0n1gA9-ubYt40Hag:-NBT0n1gA9-ubYt40Hah:brt8msh) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1939)):*
> ```Go
> // c.remainingRows with at least one element.
> func (c *checkConsistencyGenerator) maybeRefillRows(ctx context.Context) {
> 	if len(c.descs) == 0 || len(c.remainingRows) > 0 {
> ```

supernit: this line breaks my logical flow a bit, I would swap it as `len(c.remainingRows) > 0 || len(c.descs) == 0`.
Logically: do I still have things in the current `remainingRows`? no? ok, then do I have things in `c.descs` (i.e. can I refill `remainingRows`)?
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1943 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBT6GbICWOrnYIed0Ja:-NBT6GbICWOrnYIed0Jb:b-w0hcby) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1943)):*
> ```Go
> 		return
> 	}
> 	defer func(tBegin time.Time) {
> ```

nice pattern



_Code quote:_
```Go
	defer func(tBegin time.Time) {
		c.lastRefillDuration = timeutil.Since(tBegin)
	}(timeutil.Now())
```
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1958 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBT21pI5ksYUxWDbN5u:-NBT21pI5ksYUxWDbN5v:be15zgg) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1958)):*
> ```Go
> 	if err != nil {
> 		// Emit result as a row, and keep going.
> 		c.remainingRows = []roachpb.CheckConsistencyResponse_Result{
> ```

reuse the old `c.remainingRows` to avoid an allocation? (size was > 0)

```golang
c.remainingRows = append(c.remainingRows[:0], roachpb.CheckConsistencyResponse_Result{
	...
})
```

Bonus: -1 indentation :laughing: 
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1969 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBT2qBB2faug6xF8GdW:-NBT2qBB2faug6xF8GdX:b5zsf8k) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1969)):*
> ```Go
> 	}
> 
> 	// NB: this could have more than one entry, if a range split in the
> ```

Can this ever have 0 entries? If it can then iteration can end prematurely on the first block of size 0.
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1976 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBT7Wi057ld3OOnBddA:-NBT7Wi057ld3OOnBddB:bwuvoem) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1976)):*
> ```Go
> // Next is part of the tree.ValueGenerator interface.
> func (c *checkConsistencyGenerator) Next(ctx context.Context) (bool, error) {
> 	// Invariant: len(c.remainingRows) > 0.
> ```

This invariant is conditional to that this `Next` doesn't follow another `Next` that returned false (otherwise invariant does not hold, and this call panics). That's probably fine: the caller should use the type correctly.
Put differently: `Next() == false` is not idempotent.

In my original suggestion you would have a "truer" invariant though: `remainingRows[0]` is the previous entry (originally dummy), regardless of the misuse of the type.
Put differently: in my suggestion `Next() == false` is idempotent.
That's achieved by a `len > 1` condition, which you have to do anyway in some form (before or after resizing the slice). So you get this property for free.
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1978 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBTJ0SU9DfxubJdgy2S:-NBTJ0SU9DfxubJdgy2T:by7d3qy) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1978)):*
> ```Go
> 	// Invariant: len(c.remainingRows) > 0.
> 	c.remainingRows = c.remainingRows[1:]
> 	c.maybeRefillRows(ctx) // if necessary, try to repopulate remainingRows
> ```

style nit: seems unnecessary to factor out a method, it's sufficiently small and would be understandable if inlined
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1990 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBT-tiT9NwAV3pmanKa:-NBT-tiT9NwAV3pmanKb:b-5g7xgd) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1990)):*
> ```Go
> 	}
> 	return tree.Datums{
> 		tree.NewDInt(tree.DInt(c.remainingRows[0].RangeID)),
> ```

style nit: consider avoiding the copy-paste of `c.remainingRows[0]` by introducing a local variable


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
TFTR! Could you give it another pass? I left the diff in a temp commit so you don't have to figure out what changed.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-:-NBWGgwe4vq3kRLU6xt7:bpc2up2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov)

___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1837 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBT-2ATBG3DjrJHt--9:-NBWHOJv2p-5EhLB5SxW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1837)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

This sentence got corrupted, and there is no longer a `curRow` field mentioned in it, please fix.
</blockquote></details>

Done.
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1840 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBTANX8-8_SGPO5rj_d:-NBWHPL34J6MukYORN2l:bcib6nf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1840)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

style nit / promoted content:

IMO in Go brevity is a feature, e.g. in naming. I would name this as `rows` or `resps`/`results`, for example. The "remaining" bit can be inferred from code/comments, and, besides, `remainingRows` now represents only a prefix of remaining rows to be fully precise (as opposed to, previously, the full set of them), so it's better to make the name purposefully vague than not exactly precise.

Same about the duration field below. Something like `lastDur`/`requestDur`/`refillDur` would do.

[pre-existing] `consistencyChecker` (note how it stands out from other short field names) I would name as `checker`/`runner`, if not `ch` or `cc`, or just embed it, as both "consistency" and "check" are inferred from its type and the wrapping type, plus this file is all about consistency checking anyway.
</blockquote></details>

I think this is verging into the territory of personal preference. I'm not a fan of unnecessary verbosity but I do think that overly terse variable naming may prevent others from easily absorbing the code.
I happened to rewrite some of this anyway following one of your other suggestions, and brought back a separate field to house the current row, so I took the opportunity to clear up the naming a bit and as a result they are now short.
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1917 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBTMqBN1unDcRTXMVwu:-NBWSYdZEzNmiS6ailh2:bfefbvw) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1917)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

I don't think I follow this bit, but that's probably due to my limited knowledge about meta ranges for now. I trust this bit works, but just asking questions to understand it. If you think there is a simple/short way to reflect it in the code comments, feel free (e.g. link to some doc / package / file).

I gather that `meta1`/`meta2` are levels of the 2-level "tree" of meta ranges (i.e. ranges that tell us what/where all ranges are). When the whole keyspace is small, we only have one meta1, and (by the looks of it from the comment) also one meta2 (as opposed to not having meta2 at all). Correct?

Can both meta1 and meta2 split, or one of them is always a singleton? It both can split, where is the "global" root of things, i.e. is there some meta0 too?

What is `r1`, and why can we get a second copy?
</blockquote></details>

The meta addressing is a true brain teaser and I'd have to re-explain it to myself first before I could write a convincing comment :laughcry:
But here's a paper trail:
https://github.com/cockroachdb/cockroach/blob/32622e1b18030bd52529841ec1bb280a5683d5cb/pkg/upgrade/upgradecluster/cluster.go#L163-L179
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1939 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBT0n1gA9-ubYt40Hag:-NBWJfaQEIhhYMnlZI5S:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1939)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

supernit: this line breaks my logical flow a bit, I would swap it as `len(c.remainingRows) > 0 || len(c.descs) == 0`.
Logically: do I still have things in the current `remainingRows`? no? ok, then do I have things in `c.descs` (i.e. can I refill `remainingRows`)?
</blockquote></details>

Done.
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1958 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBT21pI5ksYUxWDbN5u:-NBWIJ0LCX_SVWBioCFN:b2lfp17) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1958)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

reuse the old `c.remainingRows` to avoid an allocation? (size was > 0)

```golang
c.remainingRows = append(c.remainingRows[:0], roachpb.CheckConsistencyResponse_Result{
	...
})
```

Bonus: -1 indentation :laughing: 
</blockquote></details>

This has changed now as a byproduct of other changes, but just to respond to your suggestion, I generally try to stay away from trying to be even mildly clever about memory reuse in cases like this where it has no benefit. The difference in memory usage produces zero difference here, but the chance that I'll slip in a dumb bug is always nonzero, and every future reader will spend a little bit more brain juice convincing themselves that the reuse is correct; that is not worth it. Besides, the other, common, path doesn't (can't) reuse.
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1969 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBT2qBB2faug6xF8GdW:-NBWJjeT50F7eDXArsJY:bmgc3s0) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1969)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Can this ever have 0 entries? If it can then iteration can end prematurely on the first block of size 0.
</blockquote></details>

No, a request always has at least one result.
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1976 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBT7Wi057ld3OOnBddA:-NBWIrUdB3AAYXGwH682:b2a0ut2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1976)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

This invariant is conditional to that this `Next` doesn't follow another `Next` that returned false (otherwise invariant does not hold, and this call panics). That's probably fine: the caller should use the type correctly.
Put differently: `Next() == false` is not idempotent.

In my original suggestion you would have a "truer" invariant though: `remainingRows[0]` is the previous entry (originally dummy), regardless of the misuse of the type.
Put differently: in my suggestion `Next() == false` is idempotent.
That's achieved by a `len > 1` condition, which you have to do anyway in some form (before or after resizing the slice). So you get this property for free.
</blockquote></details>

Hmm, good point, `Next` should be idempotent. I don't see anything in the contract that prevents the caller from invoking it again after a `(false, nil)` result.
Gave your idea another spin, but I think it is clearer to just bring `curRow` back (under a new, short, name :smile:). Using a slice that is "empty" when its length is one is surprising. It came out better with an extra field.
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1978 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBTJ0SU9DfxubJdgy2S:-NBWPNqz8K0VIS8SUhH_:b-j5ky15) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1978)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

style nit: seems unnecessary to factor out a method, it's sufficiently small and would be understandable if inlined
</blockquote></details>

I think it is helpful to keep around as is. Inlining the method leads to more indentation (it will converge back towards the original pattern where all of the refill code goes into the conditional `if <no more rows stashed> { ... }`. The early-return in `maybeRefillRows` now takes care of avoiding that.
The flow is also clearer with this method on the side, now you don't have to worry about how new rows are produced to see how they flow out to `Values` via `Next`.


<!-- Sent from Reviewable.io -->

pav-kv reviewed (approved):
Still LGTM, with a few optional things.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-:-NBWuMhYDqkLFqqcMKBr:b-j4d6l7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov and @tbg)

___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1840 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBTANX8-8_SGPO5rj_d:-NBWdUrHEE0Se-B9mv1r:b-p5r3iq) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1840)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I think this is verging into the territory of personal preference. I'm not a fan of unnecessary verbosity but I do think that overly terse variable naming may prevent others from easily absorbing the code.
I happened to rewrite some of this anyway following one of your other suggestions, and brought back a separate field to house the current row, so I took the opportunity to clear up the naming a bit and as a result they are now short.
</blockquote></details>

I agree, there is a balance to it. Thanks for addressing, looks stylish now :laughing: 
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1917 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBTMqBN1unDcRTXMVwu:-NBWrZCW7u_czJzWerLD:b-oho56q) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1917)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

The meta addressing is a true brain teaser and I'd have to re-explain it to myself first before I could write a convincing comment :laughcry:
But here's a paper trail:
https://github.com/cockroachdb/cockroach/blob/32622e1b18030bd52529841ec1bb280a5683d5cb/pkg/upgrade/upgradecluster/cluster.go#L163-L179
</blockquote></details>

Yeah, looks quite subtle. The code seems ok to me at this point.

Is it true that `len(desc.StartKey) == 0` iff it's `r1`, i.e. it's the first range in the global keyspace? Why do we bump the `StartKey` though? Do we intentionally skip some part of keyspace by doing so?
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1969 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBT2qBB2faug6xF8GdW:-NBWnSOR83Wk9uM6XaAX:bv14dv) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1969)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

No, a request always has at least one result.
</blockquote></details>

SGTM. Feel free to assert / return error otherwise though, if there is a non-intrusive way to do so.
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1976 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBT7Wi057ld3OOnBddA:-NBWgRxZ6lMVnKYWDPv2:buhzwui) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1976)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Hmm, good point, `Next` should be idempotent. I don't see anything in the contract that prevents the caller from invoking it again after a `(false, nil)` result.
Gave your idea another spin, but I think it is clearer to just bring `curRow` back (under a new, short, name :smile:). Using a slice that is "empty" when its length is one is surprising. It came out better with an extra field.
</blockquote></details>

Ack. Having or not having a separate `cur` is orthogonal to other aspects. It's either `cur`+`next[]` playing the role of one slice, or one slice `next[]` with 0-th element playing the role of `cur`. It just changes `len > 0` checks to `len > 1` or vice versa, but all other structural things are independently tweakable (like: in which order to do the checks to make things idempotent, or how to reduce nesting, etc). Agree that `cur`+`next[]` is slightly more human-readable.

E.g. my original proposal can be tweaked like this:

```golang
// No invariant needed. Returns false idempotently.
if len(c.next) != 0 {
	c.cur, c.next = c.next[0], c.next[1:]
	return true, nil
}
if len(c.descs) == 0 {
	return false, nil
}
// ... the unconditional refill here, no nesting, can call c.refill(ctx) or inline
c.cur, c.next = resp.Result[0], resp.Result[1:]
return true, nil
```
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1978 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBTJ0SU9DfxubJdgy2S:-NBWlm2tDlSP4iVBvwVL:b3gpyb8) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1978)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I think it is helpful to keep around as is. Inlining the method leads to more indentation (it will converge back towards the original pattern where all of the refill code goes into the conditional `if <no more rows stashed> { ... }`. The early-return in `maybeRefillRows` now takes care of avoiding that.
The flow is also clearer with this method on the side, now you don't have to worry about how new rows are produced to see how they flow out to `Values` via `Next`.
</blockquote></details>

It doesn't have to converge back to nested blocks etc, can be inlined pretty much verbatim. See another comment with the code snippet.
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1961 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBWw5GTAVxUv44SJJhr:-NBWw5GTAVxUv44SJJhs:bq51my1) ([raw file](https://github.com/cockroachdb/cockroach/blob/80c06942e77d3d2a5702e3a68c38fb6f6e292b2a/pkg/sql/sem/builtins/generator_builtins.go#L1961)):*
> ```Go
> 		ctx, desc.StartKey.AsRawKey(), desc.EndKey.AsRawKey(), c.mode,
> 	)
> 	if err != nil {
> ```

In relation to the `len(resp.Results) == 0` comment thread. If you'd like to protect from that case, could add a check like:

```golang
if err == nil && len(resp.Results) == 0 {
	err = <some-internal-error>
}
```


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87378)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov)

___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1917 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBTMqBN1unDcRTXMVwu:-NBXQGdk7IS4e6dVelk8:btj1bys) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1917)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Yeah, looks quite subtle. The code seems ok to me at this point.

Is it true that `len(desc.StartKey) == 0` iff it's `r1`, i.e. it's the first range in the global keyspace? Why do we bump the `StartKey` though? Do we intentionally skip some part of keyspace by doing so?
</blockquote></details>

Yes, range one starts with the empty key. But we have a concept of local keys, which live in a parallel plane (like the range descriptor key) and start with some byte. Check https://github.com/cockroachdb/cockroach/blob/2675c7c0481bb33eb9f43f1c2b52c152fca3ae2d/pkg/keys/doc.go. I am so glad that stuff mostly works, it is quite intricate and really shouldn't ever rear its ugly head in any usage of KV like here. Glad you're digging in! These are some of the hard non-distributed bits of CRDB.
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1976 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBT7Wi057ld3OOnBddA:-NBXRP1WEayx2zJRbVEl:bhcxj8s) ([raw file](https://github.com/cockroachdb/cockroach/blob/9663668e490b148a7b5c939db7be276b2a1650a1/pkg/sql/sem/builtins/generator_builtins.go#L1976)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Ack. Having or not having a separate `cur` is orthogonal to other aspects. It's either `cur`+`next[]` playing the role of one slice, or one slice `next[]` with 0-th element playing the role of `cur`. It just changes `len > 0` checks to `len > 1` or vice versa, but all other structural things are independently tweakable (like: in which order to do the checks to make things idempotent, or how to reduce nesting, etc). Agree that `cur`+`next[]` is slightly more human-readable.

E.g. my original proposal can be tweaked like this:

```golang
// No invariant needed. Returns false idempotently.
if len(c.next) != 0 {
	c.cur, c.next = c.next[0], c.next[1:]
	return true, nil
}
if len(c.descs) == 0 {
	return false, nil
}
// ... the unconditional refill here, no nesting, can call c.refill(ctx) or inline
c.cur, c.next = resp.Result[0], resp.Result[1:]
return true, nil
```
</blockquote></details>

Yeah, this trades the indentation for having two places where the array is sliced. Honestly, this is all comparable. I'm going to leave as is.

BTW it might be more productive, in the future, if you push a commit with your suggested refactor, then we don't need to tunnel it through a review which is a slow path for both of us. If you don't have it yet, `gh` makes this really easy: `gh pr checkout <NNNN>; <edit>; git commit -m foo; git push`. I've done this a few times before when reviewing and thought it worked out well.

Same for nits or small corrections (typos, etc), which cost a lot of time relative to the benefit they provide when they're communicated through github comments. At least I always prefer a few commits that I can then patch in if I can get them.
___
*[`pkg/sql/sem/builtins/generator_builtins.go` line 1961 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/87378#-NBWw5GTAVxUv44SJJhr:-NBXQZTz-DN6n7GZD9Tc:bad3exo) ([raw file](https://github.com/cockroachdb/cockroach/blob/80c06942e77d3d2a5702e3a68c38fb6f6e292b2a/pkg/sql/sem/builtins/generator_builtins.go#L1961)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

In relation to the `len(resp.Results) == 0` comment thread. If you'd like to protect from that case, could add a check like:

```golang
if err == nil && len(resp.Results) == 0 {
	err = <some-internal-error>
}
```
</blockquote></details>

I think this kind of assertion is taking it too far, since we'd need it like 1000x throughout the codebase. If anything we could bake it into the KV client (if it isn't there already) but either way introducing it in this PR isn't too useful.


<!-- Sent from Reviewable.io -->

tbg commented:
bors r=pavelkalinnikov
TFTR!

tbg commented:
blathers backport 22.2
